KEEP rewording (#14)
47deg/KEEP: proposals/compile-time-dependency-resolution.md

# Compile time dependency resolution

* **Type**: Design proposal
* **Author**: Raul Raja
* **Contributors**: Tomás Ruiz López, Jorge Castillo, Francesco Vasco, Claire Neveu
* **Status**: New
* **Prototype**: [initial implementation](https://github.com/arrow-kt/kotlin/pull/6)

## Summary

> The goal of this proposal is to enable **compile time dependency resolution** through extension syntax. Overall, we'd want to enable extension contract interfaces to be defined as function or class constructor arguments and enable compiler to automatically resolve and inject those instances that must be provided evidence for in one of a given set of scopes. In case of not having evidence of any of those required interfaces (program constraints), compiler would fail and provide proper error messages.

Reviewer: I would add somewhere...
```
The goal of this proposal is to bring first-class named extensions families to Kotlin.

Extension families allow us to guarantee a given data type (class, interface, etc.) satisfies behaviors (group of functions) that are decoupled from the type's inheritance hierarchy.

Extension families favor horizontal composition based on compile-time resolution between types and their extensions vs the traditional subtype style composition where users are forced to extend and implement classes and interfaces.
```

Author: Done in 535c453

## Motivation

* Support compile-time verification of program dependencies (extensions).
* Enable nested extension resolution.
* Support compile-time verification of a program correctness given behavioral constraints are raised to the interface types.
* Enable definition of polymorphic functions whose constraints can be verified at compile time in call sites.

## Description

We propose to use the existing `interface` semantics, allowing for generic definition of behaviors and their instances in the same style interfaces are defined.

```kotlin
package com.data

interface Repository<A> {
  fun loadAll(): List<A>

>   fun loadById(id: Int): A?

Reviewer: we should add an example with `fun A.save(): Unit` to show this is actually about grouping extensions too.

Author: Done in fba1db3 along with another change suggested by @pakoito 👍

}
```

The above declaration can serve as a target for implementations for any arbitrary type passed for `A`.

In the implementation below we provide evidence that there is a `Repository<User>` extension available in scope, enabling both methods defined for the given behavior to work over the `User` type. As you can see, we're enabling a new keyword here: `extension`.

```kotlin
package com.data.instances

import com.data.Repository
import com.domain.User

extension object UserRepository: Repository<User> {
  override fun loadAll(): List<User> {
    return listOf(User(25, "Bob"))
  }

  override fun loadById(id: Int): User? {
    return if (id == 25) {
      User(25, "Bob")
    } else {
      null
    }
  }
}
```

You can also provide evidence of an extension using classes:

```kotlin
package com.data.instances

import com.data.Repository
import com.domain.User

extension class UserRepository: Repository<User> {
  override fun loadAll(): List<User> {

>     return listOf(User(25, "Bob"))

Reviewer 2: For the sake of the example I'd make the list of users a map field.

Author: Done in fba1db3 along with another change suggested by @raulraja 👍

  }

  override fun loadById(id: Int): User? {
    return if (id == 25) {
      User(25, "Bob")
    } else {
      null
    }
  }
}
```

> **Extensions are named** for now, mostly with the purpose of supporting Java. We'd be fine with this narrower approach, but we'd be open to iterate that towards allowing definition through properties and anonymous classes, if there's a need for it.

Reviewer 2: This paragraph talks about the proposal as if it was just introduced, instead of laying out the follow-up work.

Author: The KEEP is already coded, so being fair it's correct. Still the wording might not be the best one, worked around it in bd1416d.

Now we've got the constraint definition (interface) and a way to provide evidence of an extension for it, we'd just need to connect the dots. Interfaces can be used to define constraints of a function or a class constructor. We use the `with` keyword for that.

```kotlin
fun <A> fetchById(id: String, with repository: Repository<A>): A? {
  return loadById(id) // Repository syntax is automatically activated inside the function scope!
}
```

As you can see, we get the constraint syntax automatically active inside the function scope, so we can call it's functions at will. That's because we consider `Repository` a constraint of our program at this point. In other words, the program cannot work without it, it's a requirement. That means the following two functions would be equivalent:

```kotlin
// Kotlin + KEEP-87
fun <A> fetchById(id: String, with repository: Repository<A>): A? {

>   return loadById(id)

Reviewer 2: Nit: loadById takes an Int parameter

Author: Fixed in 87d1c43.

}

// Regular Kotlin
fun <A> fetchById(id: String, repository: Repository<A>): A? =

>   with (semigroup) {

Reviewer 3: `with (repository)` ?

Author: Fixed in c8bf869.

  override fun loadAll(): Group<A> = with (R) {
    Group(loadAll())
  }
}
```

## Composition and chain of evidences

Constraint interface declarations and extension evidences can encode further constraints on their type parameters so that they can be composed nicely:

```kotlin
package com.data.instances

import com.data.Repository
import com.domain.User
import com.domain.Group

extension class GroupRepository<A>(with val repoA: Repository<A>) : Repository<Group<A>> {
  override fun loadAll(): List<Group<A>> {
    return listOf(Group(repoA.loadAll()))
  }

  override fun loadById(id: Int): Group<A>? {
    return Group(repoA.loadById(id))
  }
}
```

The above extension provides evidence of a `Repository<Group<A>>` as long as there is a `Repository<A>` in scope. Call site would be like:

```kotlin
fun fetchGroup<A>(with repo: GroupRepository<A>) = repo.loadAll()

fun main() {
  fetchGroup<User>() // Succeeds! There's evidence of Repository<Group<A>> and Repository<User> provided in scope.
  fetchGroup<Coin>() // Fails! There's evidence of Repository<Group<A>> but no evidence of `Repository<Coin>` available.
}
```

> We believe the encoding proposed above fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other langs that also support compile time dependency resolution such as Scala where this is done via implicits.

Reviewer 2: I'd drop the comparison to implicits here. It requires knowing how they work, and it doesn't give much. Maybe even the whole paragraph now that the proposal is already a body of work.

Author: Removed in 6ba2577.

## Language changes

* Add `with` to require evidence of extensions in both function and class/object declarations.
* Add `extension` to provide instance evidence for a given interface.

Usage of these language changes are demonstrated by the previous and below examples:

#### Class constraint

```kotlin
extension class GroupRepository<A>(with R: Repository<A>) : Repository<Group<A>> {
  /* ... */
}
```

#### Function constraint

```kotlin
fun <A> fetch(id: String, with R: Repository<A>): A = loadById(id) // function position using parameter "R"
```

> #### Extension evidence using an Object

Reviewer: Do extensions on `fun` and `val` work? We need to add those cases here as well.

Author: Not yet, I'd say. This is part of what we discussed in the call yesterday, we're just supporting named extensions through explicit objects or classes for now. Can you confirm that @truizlop? If we require that then I'll add it to the "What needs to be done?" section.

Reviewer 4: Exactly, right now we only support `extension class` and `extension object`, so if those cases need to be supported, they need to be added to the pending list. Is there any use case they cover, that is not covered with the currently supported options?

Reviewer: I believe both `fun` and `val` should be supported as valid providers of extensions.
`fun` unlike `class` allows you to provide instances without incurring in allocations.
`val` unlike object can be parametric if it's a property and it's perhaps let's interesting than `fun`.

Author: How would those look like? I imagine they could support nested provisioning too, but what do you mean by not incur in allocations? You'd need to allocate for providing an instance, no matter what. You probably mean allocations prior to be called?

Author: I've added it to the TODO list in 297e65f. Can you review it @raulraja? Thx.

```kotlin
extension object UserRepository: Repository<User> {
  override fun loadAll(): List<User> {
    return listOf(User(25, "Bob"))
  }

  override fun loadById(id: Int): User? {
    return if (id == 25) {
      User(25, "Bob")
    } else {
      null
    }
  }
}
```

#### Extension evidence using a Class

```kotlin
extension class UserRepository: Repository<User> {
  override fun loadAll(): List<User> {
    return listOf(User(25, "Bob"))
  }

  override fun loadById(id: Int): User? {
    return if (id == 25) {
      User(25, "Bob")
    } else {
      null
    }
  }
}
```

> 

Reviewer 2: Should we mention anything about compatibility with sealed, data, and inline classes here?

Author: That would require some exploration though. @truizlop do you know what's happening for those?

Reviewer 4: @pakoito do you mean adding `extension` to a sealed / data / inline class, or providing an extension to one of those? I haven't explored the former case though; is there any use case they cover, that cannot be covered with the existing support?

Reviewer 2: I mean whether they're compatible, yes.

## Extension resolution order

Classical interfaces only permit their implementation at the site of a type definition. Compiler extension resolution pattern typically relax this rule and **allow extension evidences be declared outside of the type definition**. When relaxing this rule it is important to preserve the coherency we take for granted with classical interfaces.

> For those reasons constraint interfaces must be declared in one of the following places (in strict resolution order):

Reviewer 2: `declared` is not the correct word because an argument in 1 is not a declaration.

Author: Fixed in 67c8f3e.

4. Subpackages of the package where the target type (User) to resolve is defined.
5. Subpackages of the package where the constraint interface (Repository) is defined.

All other instances are considered orphan instances and are not allowed. See [Appendix A](#Appendix-A) for a modification to this proposal that allows for orphan instances.

Additionally, a constraint extension must not conflict with any other pre-existing extension for the same constraint interface; for the purposes of checking this we use the normal resolution rules. That's what we refer as compiler "coherence".

#### 1. Arguments of the caller function

It looks into the caller function argument list for an evidence of the required extension. Here, `bothValid()` gets a `Validator<A>` passed in so whenever it needs to resolve it for the inner calls to `validate()`, it'll be able to retrieve it from its own argument list.
```kotlin
fun <A> validate(a: A, with validator: Validator<A>): Boolean = a.isValid()

fun <A> bothValid(x: A, y: A, with validator: Validator<A>): Boolean = validate(x) && validate(y)
```

#### 2. Companion object for the target type

> In case there's no evidence at the caller function level, we'll look into the companion of the target type. Let's say we have `Repository<A>` and `User` for the `A` type:

Reviewer: ```suggestion
In case there's no evidence at the caller function level, we'll look into the companion of the target type. Let's say we have an extension of `Validator<User>`:
```

```kotlin
data class User(val id: Int, val name: String) {
  companion object {
    extension class UserValidator(): Validator<User> {
      override fun User.isValid(): Boolean {
        return id > 0 && name.length > 0
      }
    }
  }
}
```
That'll be enough for resolving the extension.

#### 3. Companion object for the constraint interface we're looking for

In case there's neither evidence in the companion of the target type, we'll look in the companion of the constraint interface:

```kotlin
interface Validator<A> {
  fun A.isValid(): Boolean

  companion object {

>     extension class GroupValidator<A>(with val userValidator: Validator<User>) : Validator<Group> {

Reviewer 2: I like that we're using `Validator<User>` here when it was just defined above!

      override fun Group.isValid(): Boolean {
        for (x in users) {
          if (!x.isValid()) return false
        }
        return true
      }
    }
  }
}
```

#### 4. Subpackages of the package where the target type is defined

The next step would be to look into the subpackages of the package where the target type (`User`) is declared.

```kotlin
package com.domain.repository

import com.domain.User

extension object UserRepository: Repository<User> {
  override fun loadAll(): List<User> {
    return listOf(User(25, "Bob"))
  }

  override fun loadById(id: Int): User? {
    return if (id == 25) {
      User(25, "Bob")
    } else {
      null
    }
  }
}
```

> Here we got a `Repository<User>` defined in a subpackage of `com.domain`, where the `User` type is defined.

Reviewer 2: `com.domain` should be added to the snippet where `User` is defined.

Author: Fixed in 7792dd7.

#### 5. Subpackages of the package where the constraint interface is defined

Last place to look at would be subpackages of the package where the constraint interface is defined.

```kotlin
package com.data.instances

import com.data.Repository
import com.domain.User

extension object UserRepository: Repository<User> {

>   override fun loadAll(): List<User> {

Reviewer 2: Make the implementation use a Map instead.

Author: Done in fba1db3 along with another change suggested by @raulraja 👍

    return listOf(User(25, "Bob"))
  }

  override fun loadById(id: Int): User? {
    return if (id == 25) {
      User(25, "Bob")
    } else {
      null
    }
  }
}
```

> Here, we are resolving it from `com.data.instances`, which a subpackage of `com.data`, where our constraint `Repository` is defined.

Reviewer 2: Same as above, I'd add that to the snippet above.

Author: Repo is already defined in `package com.data`.

## Error reporting

We've got a `CallChecker` in place to report inlined errors using the context trace. That allows us to report as many errors as possible in a single compiler pass. Also provide them in two different formats:

#### Inline errors while coding (using inspections and red underline)

Whenever you're coding the checker is running and proper unresolvable extension errors can be reported within IDEA inspections.

> ![Screenshot 2019-04-05 at 12 51 11](https://user-images.githubusercontent.com/6547526/55622809-85580480-57a1-11e9-8254-0830d29593ec.png)

Reviewer: nice!

Reviewer: The actual error reporting should use the name of the types involved instead of those generic messages.

Reviewer 2: Is `scope of the caller function` correct? Shouldn't it be `parameters of the called function`? Are we resolving `this` as a potential parameter?

Author: Good catch! I've fixed KEEP errors now so they mention "Parameters of the caller function" as the first scope now.  The wording was not correct. Screenshots here have been updated also. Fixed in 3753301.

#### Errors once you hit the "compile" button:

Once you hit the "compile" button or run any compile command you'll also get those errors reported.

> ![Screenshot 2019-04-05 at 12 59 54](https://user-images.githubusercontent.com/6547526/55623259-be44a900-57a2-11e9-927e-fe0150265ba5.png)

Reviewer: Ideally all the screenshots remain in the context of the examples instead of mentioning things like `Semigroup`

Author: Fixed for both screenshots in 87eafe6.

## How to try KEEEP-87? (First approach)

- Clone [Our Kotlin fork](https://github.com/arrow-kt/kotlin) and checkout the **keep-87** branch.
- Follow the instructions on the [README](https://github.com/arrow-kt/kotlin/blob/master/ReadMe.md#build-environment-requirements) configure the necessary JVMs.
- Follow the instructions on the [README](https://github.com/arrow-kt/kotlin/blob/master/ReadMe.md#-working-with-the-project-in-intellij-idea) to open the project in IntelliJ IDEA.
- Once you have everything working, you can run a new instance of IntelliJ IDEA with the new modifications to the language by executing `./gradlew runIde`. There is also a pre-configured run configuration titled **IDEA** that does this.
- It will open a new instance of the IDE where you can create a new project and experiment with the new features of the language. You can also download [this project](https://github.com/arrow-kt/kotlin/files/3064100/Keep87Sample.zip)
) with some sample code that you can try out.

> ## How to try KEEEP-87? (Alternative approach - easier)

Reviewer: If this is the easier approach it should be listed first and the other one as the alternative.

Author: Done in 21aeaed.

- Add our Amazon s3 plugin repository as in the image.
![InstallKeepFromRepository2](https://user-images.githubusercontent.com/6547526/55884427-562e0280-5ba8-11e9-98e8-8811e8e3e8b0.png)
- Now browse for "keep87" plugin.
![InstallKeepFromRepository3](https://user-images.githubusercontent.com/6547526/55884468-67770f00-5ba8-11e9-92d6-e9cc8cc3f572.png)
- Install it.
![InstallKeepFromRepository4](https://user-images.githubusercontent.com/6547526/55884479-6b0a9600-5ba8-11e9-8a19-0eec53187fc5.png)
- Download and run [this project](https://github.com/arrow-kt/kotlin/files/3064100/Keep87Sample.zip) on that IntellIJ instance.

## What's still to be done?

### Type-side implementations

We have additional complications when you consider multi-parameter constraint interfaces.

```kotlin
package foo.repo

// I stands for the index type, A for the stored type.
interface Repository<I, A> {
   ...
}
```

```kotlin
package data.foo

data class Id(...)
extension class RepoIndexedById<A> : Repository<Id, A> {
   ...
}
```

```kotlin
package data.foo.user

data class User(...)
extension class UserRepository<I> : Repository<I, User> {
   ...
}
```

> The above instances are each defined alongside their respective type definitions and yet they clearly conflict with each other. We will also run into quandaries once we consider generic types. We can crib some prior art from Rust<sup>1</sup> to help us out here.

Reviewer 2: quandaries?

Author: "a state of perplexity or uncertainty over what to do in a difficult situation."

    return loadById(id)
  }
```

On the call site, we could use it like:

```kotlin
fetchById<User>(11829) // compiles since we got evidence of a `Repository<User>` in scope.
fetchById<Coin>(12398) // does not compile: No `Repository<Coin>` evidence defined in scope!
```

Functions with extension parameters can be passed all values, or extension ones can be omitted and let the compiler resolve the suitable extensions for them. That makes the approach really similar to how default arguments work in Kotlin.

```kotlin
fetchById<User>(11829) // compiles since we got evidence of a `Repository<User>` in scope.
fetchById<User>(11829, UserRepository()) // you can provide it manually.
```

When `with` is used in class constructors, it is important to **add val to extension class fields** to make sure they are accessible in the scope of the class. Here, the `with` keyword adds the value to the scope of every method in the class. In this scenario, the following classes would be equivalent:

```kotlin
data class Group<A>(val values: List<A>)

// Kotlin + KEEP-87
extension class GroupRepository<A>(with val R: Repository<A>) : Repository<Group<A>> {
  override fun loadAll(): Group<A> =

>     Group(R.loadAll())

Reviewer 5: Looks good.

}

// Regular Kotlin
class GroupRepository<A>(val R: Repository<A>) : Repository<Group<A>> {
  override fun loadAll(): Group<A> = with (R) {
    Group(loadAll())
  }
}
```

## Composition and chain of evidences

Constraint interface declarations and extension evidences can encode further constraints on their type parameters so that they can be composed nicely:

```kotlin
package com.data.instances

import com.data.Repository
import com.domain.User
import com.domain.Group

extension class GroupRepository<A>(with val repoA: Repository<A>) : Repository<Group<A>> {
  override fun loadAll(): Map<Int, Group<A>> {
    return repoA.loadAll().mapValues { Group(it.value) }
  }

  override fun loadById(id: Int): Group<A>? {
    return repoA.loadById(id)?.let { Group(it) }
  }

  override fun Group<A>.save() {
    this.items.map { repoA.run { it.save() } }
  }
}
```

The above extension provides evidence of a `Repository<Group<A>>` as long as there is a `Repository<A>` in scope. Call site would be like:

```kotlin

> fun fetchGroup<A>(with repo: GroupRepository<A>) = repo.loadAll()

Author: Yep, fixed in `951ea3d`.

    return storedUsers[id]
  }

  override fun User.save() {
    storedUsers[this.id] = this
  }
}
```

#### Extension evidence using a Class

```kotlin
extension class UserRepository: Repository<User> {

  val storedUsers: MutableMap<Int, User> = mutableMapOf() // e.g. users stored in a DB

  override fun loadAll(): Map<Int, User> {
    return storedUsers
  }

  override fun loadById(id: Int): User? {
    return storedUsers[id]
  }

  override fun User.save() {
    storedUsers[this.id] = this
  }
}
```

## Extension resolution order

Classical interfaces only permit their implementation at the site of a type definition. Compiler extension resolution pattern typically relax this rule and **allow extension evidences be declared outside of the type definition**. When relaxing this rule it is important to preserve the coherency we take for granted with classical interfaces.

For those reasons constraint interfaces must be provided in one of the following scopes (in strict resolution order):

1. Arguments of the caller function.
2. Companion object for the target type (User).
3. Companion object for the constraint interface we're looking for (Repository).

> 4. Subpackages of the package where the target type (User) to resolve is defined.

Reviewer 5: Subpackages. So including across modules?

Reviewer 4: Nope. When instances are defined in subpackages of type or type class, they need to be marked as internal.

Reviewer 5: Should we explicitly add that remark here?

Author: Yep definitely. Sorry, resolved this one too early. I'll add that now.

Author: Fixed in 24b9dfe.

### Using extensions in inlined lambdas

Inlined functions get into trouble when it comes to capture resolved extensions.[The problem is described here](https://github.com/arrow-kt/kotlin/issues/14).

### Function and property extension providers

Ideally we'd enable users to provide extensions also using `val` and `fun`. They'd look similar to:

```kotlin
// Simple fun extension provisioning
extension fun userRepository(): Repository<User> = object : Repository<User>() {
	/* ... */
}

// Chained fun extension provisioning (would require both to resolve).
extension fun userValidator(): Validator<User> = UserValidator()
extension fun userRepository(with validator: Validator<User>) : Repository<User> = UserRepository(validator)

// Simple extension provisioning
extension val userRepository: Repository<User> = UserRepository()

// Chained extension provisioning (would require both to resolve).
extension val userValidator: Validator<User> = UserValidator()
extension val userRepository: Repository<User> = UserRepository(userValidator)
```

> In chained extension provisioning, all chained extensions would be required from the caller scope, but not necessarily all provided in the same resolution scope. E.g: `Repository<User>` could be provided in a different resolution scope than `Validator<User>`, and the program would still compile successfully as long as both are available.

Author: I might not be using the proper term here (open for suggestions) but I'm referring to nesting extension resolutions up to N arbitrary levels. An extension can require a constraint that is provided somewhere else (e.g: Dagger providers). That's something already working for classes and objects, so we would also be needed for funs and vals. Just one pass is needed AFAIK.

About "could be provided in a different scope" I mean you can compose dependency resolution from N different resolution scopes as in the ordered steps (scopes) we look at for resolution described on this keep. On this example, `Repository<User>` could be defined in the target type companion and `Validator<User>` in the type class subpackages (just a random example), and it should suffice, since both are properly resolved.

> In chained extension provisioning, all chained extensions would be required from the caller scope, but not necessarily all provided in the same resolution scope. E.g: `Repository<User>` could be provided in a different resolution scope than `Validator<User>`, and the program would still compile successfully as long as both are available.

Author: Right, switched to `nested` in 83a1a94.

## What's still to be done?

### Instance resolution based on inheritance

> Some scenarios are not covered yet given some knowledge we are lacking about how subtyping resolution rules are coded in Kotlin compiler. The different scenarios would be required for a fully working compile time extension resolution feature, and [are described in detail here](https://github.com/arrow-kt/kotlin/issues/15).

Reviewer 5: Sidenote: Helios has a nice example Scenario 2 where `Eq<Json>` uses different typed `Eq`s to be able to compare numbers.
